6176679: Application freezes when copying an animated gif image to the system clipboard

rajamah opened the pull request:
**Problem:**

On pressing the Copy button we keep waiting in AWT-EventQueue thread in reconstruct() function as we detect that we have missing information for the animated image since we are copying single frame at a time.
Due to this infinite wait the application freezes.

**Proposed Fix:**

There are conditions in the reconstruct() function that avoid entering the wait() code if we have some error reading the image , etc. So, I added the condition to avoid entering the wait() code if we are copying single frame at a time. This sounded logical to me since if we have incomplete information(single frame) about the animated image we shouldn’t keep waiting, as it leads to infinite wait. 
After this change I see the GIF image being correctly copied and animated.


**Testing:**

Added a test for this (bug6176679.java) and tested locally on my Windows machine and on mach5.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-6176679](https://bugs.openjdk.org/browse/JDK-6176679): Application freezes when copying an animated gif image to the system clipboard


### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Dmitry Markov](https://openjdk.org/census#dmarkov) (@dmarkov20 - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13414/head:pull/13414` \
`$ git checkout pull/13414`

Update a local copy of the PR: \
`$ git checkout pull/13414` \
`$ git pull https://git.openjdk.org/jdk.git pull/13414/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13414`

View PR using the GUI difftool: \
`$ git pr show -t 13414`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13414.diff">https://git.openjdk.org/jdk/pull/13414.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13414#issuecomment-1503903219)

jayathirthrao reviewed (changes requested):
Instead of using .gif image as input, its better if we can create ImageStream as we do in some of animation tests under test/jdk/javax/imageio/plugins/gif.

aivanov-jdk commented:
> Instead of using .gif image as input, its better if we can create ImageStream as we do in some of animation tests under test/jdk/javax/imageio/plugins/gif.

@jayathirthrao Could you elaborate, please? How is `ImageStream` better than reading from a file directly?
Do you mean that it's better to store the GIF as a byte array or base64-encoded string so that there's no external .gif file? Do you mean generating the GIF by code on the fly? Anything else?

aivanov-jdk commented on test/jdk/javax/imageio/plugins/gif/6176679/bug6176679.java:
```diff
@@ -0,0 +1,184 @@
+        Panel panel = new Panel(new GridLayout(1, 1));
+        panel.add(canvas);
+        frame.add(panel);
+        img = frame.getToolkit().getImage(System.getProperty("test.src", ".") + File.separator + FILENAME);
```
Could you wrap this long line, please?

The 80-column limit isn't strictly followed, yet it's better to fit into 100-column limit because the code becomes harder to read.

aivanov-jdk commented on src/java.desktop/share/classes/sun/awt/image/ImageRepresentation.java:
```diff
@@ -109,7 +109,7 @@ public synchronized void reconstruct(int flags) {
         }
         int missinginfo = flags & ~availinfo;
-        if ((availinfo & ImageObserver.ERROR) == 0 && missinginfo != 0) {
+        if ((availinfo & (ImageObserver.ERROR |  ImageObserver.FRAMEBITS) ) == 0 && missinginfo != 0) {
```
```suggestion
        if ((availinfo & (ImageObserver.ERROR | ImageObserver.FRAMEBITS)) == 0
            && missinginfo != 0) {
```
The space between the closing parentheses is redundant as well as the extra space after `|` operator. The line could be wrapped to fit into 80-column limit.

aivanov-jdk commented on src/java.desktop/share/classes/sun/awt/image/ImageRepresentation.java:
Please update the copyright year to 2023.

jayathirthrao commented:
> > Instead of using .gif image as input, its better if we can create ImageStream as we do in some of animation tests under test/jdk/javax/imageio/plugins/gif.
> 
> @jayathirthrao Could you elaborate, please? How is `ImageStream` better than reading from a file directly? Do you mean that it's better to store the GIF as a byte array or base64-encoded string so that there's no external .gif file? Do you mean generating the GIF by code on the fly? Anything else?

Yes i mean generating GIF by code and not using .gif file as input.

rajamah commented:
> 
Will the generated GIF file be animated ?, because this only happens with animated GIFs. 
And why is it better than having the GIF file itself ?

jayathirthrao commented:
> > 
> 
> Will the generated GIF file be animated ?, because this only happens with animated GIFs. And why is it better than having the GIF file itself ?

As previously mentioned we have animation tests for GIF under test/jdk/javax/imageio/plugins/gif which don't use GIF file input.
We try to avoid adding new image file to JDK repo whenever we have some ImageIO test because it increases the size, so its better to use stream or create and delete temp file in regression tests.

aivanov-jdk commented:
> > Yes i mean generating GIF by code and not using .gif file as input.
>
> Will the generated GIF file be animated ?, because this only happens with animated GIFs. And why is it better than having the GIF file itself ?

It should if you create several frames.

There's always an option to create another GIF image, smaller, and use its byte or base64 representation as the source.

> We try to avoid adding new image file to JDK repo whenever we have some ImageIO test because it increases the size, so its better to use stream or create and delete temp file in regression tests.

Okay, now it's clear.

rajamah commented:
Made changes requested by the reviewers. Now we read animated gif image from byte array.
Thanks @aivanov-jdk for the help with this.

rajamah commented:
> 



> _Mailing list message from [Alan Snyder](mailto:javalists@cbfiddle.com) on [client-libs-dev](mailto:client-libs-dev@mail.openjdk.org):_
> 
> > On Apr 14, 2023, at 3:48 PM, Rajat Mahajan <rmahajan at openjdk.org> wrote:
> > > On pressing the Copy button we keep waiting in AWT-EventQueue thread in reconstruct() function as we detect that we have missing information for the animated image since we are copying single frame at a time.
> 
> Is it really the case that multiple frames are copied one at a time?
> 
> I don?t see that, and don?t see how that could work.
> 
> I?m not saying it is wrong to copy only the first frame, as that is the only reasonable option if the goal is to transfer a BufferedImage.
> 
> I assume there is no Java representation of a multi-frame image that could be copied?
> 
> -------------- next part -------------- An HTML attachment was scrubbed... URL: <https://mail.openjdk.org/pipermail/client-libs-dev/attachments/20230414/582a7883/attachment.htm>

Yes I see that the frame is copied one at a time. When I debug and step in reconstruct() is called per frame.
I am not aware of Java representation of a multi-frame image.

aivanov-jdk commented on src/java.desktop/share/classes/sun/awt/image/ImageRepresentation.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1995, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018, 2023 Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 1995, 2023, Oracle and/or its affiliates. All rights reserved.
```
You always update the second year leaving the first one intact. The comma after the year should remain.

aivanov-jdk commented on src/java.desktop/share/classes/sun/awt/image/ImageRepresentation.java:
```diff
@@ -109,7 +109,8 @@ public synchronized void reconstruct(int flags) {
         }
         int missinginfo = flags & ~availinfo;
-        if ((availinfo & ImageObserver.ERROR) == 0 && missinginfo != 0) {
+        if ((availinfo & (ImageObserver.ERROR | ImageObserver.FRAMEBITS)) == 0
+             && missinginfo != 0) {
```
```suggestion
        if ((availinfo & (ImageObserver.ERROR | ImageObserver.FRAMEBITS)) == 0
            && missinginfo != 0) {
```
Nit: the wrapped line should be aligned to the opening parenthesis of the `if`, otherwise it looks as if the wrapped condition is part of the condition in the parentheses but it is not.

aivanov-jdk commented on test/jdk/javax/imageio/plugins/gif/6176679/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,168 @@
+            (byte) 0x00, (byte) 0x00, (byte) 0x02, (byte) 0x04, (byte) 0x8c,
+            (byte) 0x8f, (byte) 0x19, (byte) 0x05, (byte) 0x00, (byte) 0x3b
+    };
+    private void createGUI() {
```
Please add a blank line between methods, fields and methods, between class declarations.

A blank line between unrelated fields is also a good idea.

aivanov-jdk commented on test/jdk/javax/imageio/plugins/gif/6176679/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,168 @@
+    private void copyImage() {
+        Clipboard sys = Toolkit.getDefaultToolkit().getSystemClipboard();
+        sys.setContents(new MyTransferable(img), null);
+
```
The blank line in the end of the method is redundant.

aivanov-jdk commented on test/jdk/javax/imageio/plugins/gif/6176679/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,168 @@
+            latch.countDown();
+        });
+
+        if (!latch.await(TIMEOUT, MILLISECONDS)) {
+            throw new RuntimeException("Image copying taking too long.");
+        }
+        else
+        {
+            EventQueue.invokeAndWait(CopyAnimatedGIFTest::dispose);
+        }
+
+    }
```
```suggestion
        if (!latch.await(TIMEOUT, MILLISECONDS)) {
            throw new RuntimeException("Image copying taking too long.");
        } else {
            EventQueue.invokeAndWait(CopyAnimatedGIFTest::dispose);
        }
    }
```
Let's follow Java Coding Style conventions.

I can't help [raising the question](https://github.com/openjdk/jdk/pull/13414#discussion_r1164559765) again:
> I think it also deserves a comment why the frame is disposed of only when the test doesn't fail. Perhaps, we can drop disposing altogether and let jtreg do the cleanup.

aivanov-jdk commented on test/jdk/javax/imageio/plugins/gif/6176679/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,168 @@
+            frame = null;
+        }
+    }
+    private static class imgCanvas extends Canvas {
```
Class names start with a capital letter, method names and fields start with a lower-case letter.

> Is `ImageCanvas` a better name?

Other points from [this comment](https://github.com/openjdk/jdk/pull/13414#discussion_r1164566145) have been addressed.

aivanov-jdk commented on test/jdk/javax/imageio/plugins/gif/6176679/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,168 @@
+        }
+    }
+
+}
```
Please add a final blank line.

aivanov-jdk commented on test/jdk/javax/imageio/plugins/gif/6176679/CopyAnimatedGIFTest.java:
I [still](https://github.com/openjdk/jdk/pull/13414#discussion_r1164573562) think the test should not be in `test/jdk/javax/imageio/plugins/gif` because it does **not** test the GIF Image IO plugin. A better location is under `test/java/awt/Clipboard`.

Now that there's no external GIF image file, the bugid directory has become unnecessary. We're striving to keep test structure flatter without creating a new folder for each test as it used to be done previously.

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,161 @@
+        frame.setVisible(true);
+    }
+
+    private void copyImage() {
+        Clipboard sys = Toolkit.getDefaultToolkit().getSystemClipboard();
+        sys.setContents(new MyTransferable(img), null);
+
+    }
```
```suggestion
    private void copyImage() {
        Clipboard sys = Toolkit.getDefaultToolkit().getSystemClipboard();
        sys.setContents(new MyTransferable(img), null);
    }
```
Let's unnecessary blank line.

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,161 @@
+        public void paint(Graphics g) {
+            g.drawImage(img, 0, 0, getSize().width, getSize().height, this);
+        }
+    }
+    private static class MyTransferable implements Transferable {
```
```suggestion
    }

    private static class MyTransferable implements Transferable {
```
A blank line between class declarations.

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,161 @@
+        }
+    }
+
+}
+
```
```suggestion
}
```
Now there are two blank lines in the end of the file for some reason. The thing is, GitHub and `git diff` don't usually show the blank line in the end of the file.

It's not too important, however, it may create noise when someone else updates the file and IDE removes the extra blank lines or adds ones if it's missing.

rajamah commented:
> _Mailing list message from [Alan Snyder](mailto:javalists@cbfiddle.com) on [client-libs-dev](mailto:client-libs-dev@mail.openjdk.org):_
> 
> I was hoping to be able to run your test program, but I cannot because on macOS there is no support for transferring a native image.
> 
> It also appears that I cannot simulate the transfer code because that code depends upon the reconstruct method, which is not public.
> 
> Just looking at the code, I think it is not correct.
> 
> It looks to me that the ImageRepresentation.drawToBufImage method is called only once, and it will do nothing unless ALLBITS or FRAMEBITS is true. I do not see any loop that would call it multiple times. (If I am correct about this, then only the first frame is drawn into the buffered image, which makes much more sense than drawing every frame into the buffered image. One might ask, however, what a user is trying to accomplish by copying an animated image to the system clipboard?)
> 
> So, it works if reconstruct() waits until ALLBITS or FRAMEBITS is true. But your change does not test for FRAMEBITS in the loop, so it will only work if FRAMEBITS is already true when reconstruct() is called.
> 
> How can that happen?
> 
> I think it works because you are displaying the image on the screen, and the animation code is reading frames.
> 
> I suggest trying the test without displaying the image.

Thanks for your suggestions. I have modified the code so it works for both cases where image is and is not displayed on screen.
I have modified the test so it tests for both these cases.

> Also, your test would better if it retrieved the native image from the system clipboard to verify that it was successfully transferred. You would need to clear the system clipboard before running the test to be sure you are not retrieving something from a previous test run.

I am not sure what you mean by native image. Could you please clarify ?

rajamah commented:
> _Mailing list message from [Alan Snyder](mailto:javalists@cbfiddle.com) on [client-libs-dev](mailto:client-libs-dev@mail.openjdk.org):_
> 
> Your changes look reasonable.
> 
> I have a suggestion. Unless IDEA is lying to me, all uses of reconstruct() pass ALLBITS as the parameter.
> 
> The net effect is to wait for any of the three terminating conditions, ALLBITS, FRAMEBITS, and ERROR.
> 
> My suggestion would be to eliminate the parameter, add ALLBITS to the test in the loop, and rename the method waitForTermination() or something like that.
> 
> (The existing method can hang if ALLBITS is not true in the parameter, which makes the method unnecessarily fragile.)
> 
> > I am not sure what you mean by native image. Could you please clarify ?
> 
> I think I was mistaken. I tried again and was able to run your test.
> 
> -------------- next part -------------- An HTML attachment was scrubbed... URL: <https://mail.openjdk.org/pipermail/client-libs-dev/attachments/20230430/3e4de822/attachment-0001.htm>

Thanks Alan for this suggestion. I have created a new JBS bug [JDK-8307183](https://bugs.openjdk.org/browse/JDK-8307183) for refactoring this code as I thought it is not in scope of this bug and would do it as a follow up change to this one.

aivanov-jdk commented on src/java.desktop/share/classes/sun/awt/image/ImageRepresentation.java:
```diff
@@ -114,7 +114,8 @@ public synchronized void reconstruct(int flags) {
                 startProduction();
                 missinginfo = flags & ~availinfo;
-                while ((availinfo & ImageObserver.ERROR) == 0 &&
+                while ((availinfo &
+                        (ImageObserver.ERROR | ImageObserver.FRAMEBITS)) == 0 &&
                        missinginfo != 0)
```
In my opinion, such formatting looks confusing, keeping both ERROR and FRAMEBITS on the same line improves readability even though it doesn't fit 80-column limit.
```suggestion
                while ((availinfo & (ImageObserver.ERROR | ImageObserver.FRAMEBITS)) == 0
                       && missinginfo != 0) 
```

Or alternatively, wrap `FRAMEBITS`:
```suggestion
                while ((availinfo & (ImageObserver.ERROR
                                     | ImageObserver.FRAMEBITS)) == 0
                       && missinginfo != 0)
```

rajamah replied:
I agree the single line is one is more readable, was just following  the 80 column limit you told me before :-)

aivanov-jdk commented on src/java.desktop/share/classes/sun/awt/image/ImageRepresentation.java:
Could you remove the unused imports, please?

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    private void createGUI() {
+        img = Toolkit.getDefaultToolkit().createImage(imageData);
```
`getImage` to avoid code duplication?

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,180 @@
+        sys.setContents(new MyTransferable(img), null);
+    }
+
+    private void runTest(boolean isImageDisplayed) throws AWTException, InterruptedException, InvocationTargetException {
```
The line is too long.
```suggestion
    private void runTest(boolean isImageDisplayed) throws Exception {
```
We don't really care which exceptions the test method throws.

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,180 @@
+
+        if (isImageDisplayed) {
+            EventQueue.invokeAndWait(test::createGUI);
+        }
+        else {
```
```suggestion
        } else {
```
In Java code style, you keep both the closing brace, the `else` keyword and the opening brace on the same line.

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,180 @@
+
+public class CopyAnimatedGIFTest {
+    private static final long TIMEOUT = 10000;
+    private static final CountDownLatch latch = new CountDownLatch(1);
```
You use the latch twice but initialise only once — the second test will never fail. You have to create it each time.

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    private static class ImageCanvas extends Canvas {
+        private final Image img;
+        public ImageCanvas(Image img) {
```
```suggestion
        private final Image img;

        public ImageCanvas(Image img) {
```

aivanov-jdk commented on test/jdk/java/awt/Clipboard/CopyAnimatedGIFTest.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    private void runTest(boolean isImageDisplayed) throws Exception {
+        Robot robot = new Robot();
+
+        if (isImageDisplayed) {
+            EventQueue.invokeAndWait(this::createGUI);
+        }
+
+        robot.waitForIdle();
+        robot.delay(1000);
```
Robot is needed only when UI is displayed, so it could be moved inside the `if (isImageDisplayed)` block.

aivanov-jdk reviewed (approved):
Looks good to me now.

You have to address the trailing whitespace error too.

rajamah commented:
/integrate

aivanov-jdk commented:
/sponsor
